keps/sig-cli/2953-kustomize-plugin-graduation/README.md

```diff
@@ -273,25 +275,23 @@ Optionally, it should be possible for a user to compile a bespoke distribution o
-
-For plugins distributed as containers, the `--enable-container-plugins` flag will need to be provided. The containers will continue to be denied disk and network access by default. To enable access, the existing network/mount restriction flags will be retained but renamed to clarify what they govern:  `--plugin-container-networks=[STRING]` `--plugin-container-mounts=[STRING]`.
-
-For plugins distributed as executables (typically committed in git along with the Kustomization), every provider the Kustomization needs must be explicitly named in a new flag:  `--trust-embedded-exec-plugins=base/plugin/transformer.sh,overlay/scripts/gen.sh`. By using this flag, the user demonstrates they are aware of what will be executed and accept the risk of running each one. To make the list less painful to construct, the error message thrown when this list is not provided or is incomplete will inform the user of which are missing.
+To mitigate the inconvenience to users who tend to work with a specific set of catalogs across many Kustomizations, we will introduce a Kustomize user preferences file in which a list of catalogs to be trusted by default can be stored. This list of trusted catalogs should support wildcard specification of catalog sources, e.g. `github.com/my-co/krm-functions/catalogs/*` to trust all catalogs published in or below that location. The contents of the preferences file will itself be properly typed KRM object, extensible to other use cases.
 
-**Additional constraints for exec providers**
+**Additional constraints for local exec providers**
 
-The location of exec providers that are local will be subject Kustomize's load restrictor, i.e. the LoadRestrictionsRootOnly policy will apply to them by default (it does not today). If this is deemed insufficient, we could take it a step further and force the LoadRestrictionsRootOnly policy even when a different loader policy is provided by flag.
+The location of exec providers that are local will be constrained as follows:
+- We will stop using `exec.LookPath` to find plugins. Local plugin locations will always need to be explicit, to guarantee that what Catalog is allowing and what Kustomize invokes are indeed the same thing.
+- Relative paths will be subject Kustomize's load restrictor, i.e. the LoadRestrictionsRootOnly policy will apply to them by default (it does not today).
+- Absolute paths will be permitted, and will be subjected to a new load restriction rule that requires them to be within a specific root, such as `$XGD_CONFIG_HOME/kustomize/plugins`. This is needed to support airgap use cases, as discussed in https://github.com/kubernetes-sigs/kustomize/issues/4288.
+  + This also provides a migration path for legacy exec plugins, which already exist in such a location. Kustomize should be able to derive Catalogs for them automatically in cases where they are already configured declaratively (i.e. do not require bespoke args).
```

Comment:
I agree that end users of a platform should not need to know where tools are installed for them. In this scenario, it sounds like we have up to three actors: the actor who runs the gitops platform that includes Kustomize, the actor who wants to run Kustomize for their app, and the actor who actually wrote a particular Kustomization. If I understand correctly, it's the first one that installs the plugins. If so, it seems to me that they could provide a Catalog of what they've installed, such that users wouldn't even need to reference the binary/container in their config anymore. This could be done by compiling the Catalog into their version of Kustomize, by including a flag with that Catalog in Kustomize invocations by default, and/or by exposing that Catalog to users through configuration. Users could configure additional catalogs to trust in their individual invocations, if they're able to install custom plugins as well. On the contrary, if allowing arbitrary code execution like that would be undesirable, the new system should allow the platform maintainer to compile versions of Kustomize that _only_ trust the built-in catalog, and do not allow additional plugins to be trusted at runtime. Do those options sound reasonable?

Reply:
@KnVerey that sounds good. Thank you for your help!

As long as the GitOps platform has the ability to call Kustomize (through the Go API and/or the CLI) with only allowing the single trusted catalog, that would work for my use-case.

To make sure I understand this correctly, would the GitOps platform be able to just provide the `--trusted-catalog` argument to the local catalog when executing Kustomize using the user's kustomization? The user's kustomization would then only need to provide the plugin's configuration YAML like they do today for exec plugins. Is that right?

Reply:
> As long as the GitOps platform has the ability to call Kustomize (through the Go API and/or the CLI) with only allowing the single trusted catalog, that would work for my use-case.

I've added a description of the option to disable adding catalogs at runtime in my latest commit, so we can remember that/why that option is needed.

> To make sure I understand this correctly, would the GitOps platform be able to just provide the --trusted-catalog argument to the local catalog when executing Kustomize using the user's kustomization? The user's kustomization would then only need to provide the plugin's configuration YAML like they do today for exec plugins. Is that right?

You could do it that way, yes, but then end users would be able to add the flag as well if I've understood the setup correctly. To prevent that, you'd need to build a version of Kustomize that trusts your plugins only, automatically. I've added a story about this scenario to the KEP in my last commit. Either way, yes, the end user plugin configuration will just be the resource as in the legacy exec style today.

Reply:
@KnVerey that sounds good!

In the use-case of a GitOps platform, the user wouldn't typically specify how kustomize is called. Only the kustomization that they want build. As long as additional trusted catalogs cannot be added via the kustomization file(s), then I think runtime flags that the GitOps platform always uses when executing kustomize would suffice.